README.md
# LifxApi [![Build Status](https://travis-ci.org/cyclotron3k/lifx_api.svg?branch=master)](https://travis-ci.org/cyclotron3k/lifx_api)

A Ruby client for the LIFX API.

This provides access to the [LIFX HTTP API](https://api.developer.lifx.com/), so it can control your lights from anywhere in the world. It does not implement the [LAN API](https://lan.developer.lifx.com/).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'lifx_api'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install lifx_api

## Usage

You'll need to have some LIFX bulbs already setup and configured. Then you will need to obtain an `access_token` from the [LIFX website](https://cloud.lifx.com/sign_in).

```ruby
require 'lifx_api'

access_token = "a1b7349df2a...e213"
client = LifxApi.new access_token
lights = client.list_lights
exit() if lights.count == 0
light_id = lights.first[:id]

client.toggle_power selector: 'all'

client.toggle_power selector: "id:#{light_id}"
```

## Endpoints and parameters

See [LIFX HTTP API](https://api.developer.lifx.com/) for the official documentation.

### list_lights

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.

### set_state

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:power` | string |  |  | The power state you want to set on the selector. Must be `on` or `off`.
`:color` | color |  |  | The color to set the light to. See [color](https://api.developer.lifx.com/docs/colors) documentation on the LIFX website.
`:brightness` | numeric |  |  | The brightness level from 0.0 to 1.0. Overrides any brightness set in color (if any).
`:duration` | numeric |  | 1.0 | How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)
`:infrared` | numeric |  |  | The maximum brightness of the infrared channel from `0.0` to `1.0`.
`:fast` | boolean |  | false | Execute the query fast, without initial state checks and wait for no results.

### set_states

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:states` | array of hashes | &#10004; |  | Array of state hashes as per `#set_state`. No more than 50 entries.
`:defaults` | hash |  |  | Default values to use when not specified in each `states[]` hash.
`:fast` | boolean |  | false | Execute the query fast, without initial state checks and wait for no results.

### state_delta

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:power` | string |  |  | The power state you want to set on the selector. Must be `on` or `off`.
`:duration` | numeric |  | 1.0 | How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)
`:infrared` | numeric |  |  | The maximum brightness of the infrared channel.
`:hue` | numeric |  |  | Rotate the hue by this angle in degrees. Range: `-360.0` - `360.0` degrees.
`:saturation` | numeric |  |  | Change the saturation by this additive amount; the resulting saturation is clipped to `[0, 1]`.
`:brightness` | numeric |  |  | Change the brightness by this additive amount; the resulting brightness is clipped to `[0, 1]`.
`:kelvin` | numeric |  |  | Change the kelvin by this additive amount; the resulting kelvin is clipped to `[2500, 9000]`.

### stage_delta
[DEPRECATED] `#stage_delta` is deprecated, please use `#state_delta` instead

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:power` | string |  |  | The power state you want to set on the selector. Must be `on` or `off`.
`:duration` | numeric |  | 1.0 | How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)
`:infrared` | numeric |  |  | The maximum brightness of the infrared channel.
`:hue` | numeric |  |  | Rotate the hue by this angle in degrees. Range: `-360.0` - `360.0` degrees.
`:saturation` | numeric |  |  | Change the saturation by this additive amount; the resulting saturation is clipped to `[0, 1]`.
`:brightness` | numeric |  |  | Change the brightness by this additive amount; the resulting brightness is clipped to `[0, 1]`.
`:kelvin` | numeric |  |  | Change the kelvin by this additive amount; the resulting kelvin is clipped to `[2500, 9000]`.

### toggle_power

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:duration` | numeric |  | 1.0 | The time is seconds to spend perfoming the power toggle.

### breathe_effect

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:color` | color | &#10004; |  | The color to use for the breathe effect. See [color](https://api.developer.lifx.com/docs/colors) documentation on the LIFX website.
`:from_color` | string |  | current bulb color | The color to start the effect from. If this parameter is omitted then the color the bulb is currently set to is used instead.
`:period` | numeric |  | 1.0 | The time in seconds for one cyles of the effect.
`:cycles` | numeric |  | 1.0 | The number of times to repeat the effect.
`:persist` | boolean |  | false | If `false` set the light back to its previous value when effect ends, if true leave the last effect color.
`:power_on` | boolean |  | true | If `true`, turn the bulb on if it is not already on.
`:peak` | numeric |  | 0.5 | Defines where in a period the target color is at its maximum. Minimum `0.0`, maximum `1.0`.

### move_effect

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:direction` | string |  | forward | Move direction, can be `forward` or `backward`.
`:period` | numeric |  | 1.0 | The time in seconds for one cyles of the effect.
`:cycles` | numeric |  | infinite | The number of times to move the pattern across the device. Special cases are `0` to switch the effect off, and unspecified to continue indefinitely.
`:power_on` | boolean |  | true | Switch any selected device that is off to on before performing the effect.

### morph_effect

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:period` | numeric |  | 5.0 | This controls how quickly the morph runs. It is measured in seconds. A lower number means the animation is faster.
`:duration` | numeric |  | infinite | How long the animation lasts for in seconds. Not specifying a duration makes the animation never stop. Specifying `0` makes the animation stop. Note that there is a known bug where the tile remains in the animation once it has completed if duration is nonzero.
`:palette` | array of colors |  | 7 colours across the spectrum | You can control the colors in the animation by specifying a list of color specifiers. For example `["red", "hue:100 saturation:1"]`.
`:power_on` | boolean |  | true | Switch any selected device that is off to on before performing the effect.

### flame_effect

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:period` | numeric |  | 5.0 | This controls how quickly the flame runs. It is measured in seconds. A lower number means the animation is faster.
`:duration` | numeric |  | infinite | How long the animation lasts for in seconds. Not specifying a duration makes the animation never stop. Specifying `0` makes the animation stop. Note that there is a known bug where the tile remains in the animation once it has completed if duration is nonzero.
`:power_on` | boolean |  | true | Switch any selected device that is off to on before performing the effect.

### pulse_effect

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:color` | string | &#10004; |  | The color to use for the pulse effect. See [color](https://api.developer.lifx.com/docs/colors) documentation on the LIFX website.
`:from_color` | string |  | current bulb color | The color to start the effect from. If this parameter is omitted then the color the bulb is currently set to is used instead.
`:period` | numeric |  | 1.0 | The time in seconds for one cyles of the effect.
`:cycles` | numeric |  | 1.0 | The number of times to repeat the effect.
`:persist` | boolean |  | false | If `false` set the light back to its previous value when effect ends, if true leave the last effect color.
`:power_on` | boolean |  | true | If `true`, turn the bulb on if it is not already on.

### effects_off

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:power_off` | boolean |  | false | If `true`, the devices will also be turned off.

### cycle

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:selector` | selector |  | all | See [selector](https://api.developer.lifx.com/docs/selectors) documentation on the LIFX website.
`:states` | array of hashes | &#10004; |  | Array of state hashes as per `#set_state`. Must have 2 to 5 entries.
`:defaults` | hash |  |  | Default values to use when not specified in each `states[]` object.
`:direction` | direction |  | forward | Direction in which to cycle through the list. Can be `forward` or `backward`.

### list_scenes

*No parameters*

### activate_scene

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:scene_uuid` | uuid | &#10004; |  | The UUID for the scene you wish to activate
`:duration` | numeric |  | 1.0 | The time in seconds to spend performing the scene transition.
`:ignore` | array of strings |  |  | Any of `power`, `infrared`, `duration`, `intensity`, `hue`, `saturation`, `brightness` or `kelvin`, specifying that these properties should not be changed on devices when applying the scene.
`:overrides` | hash |  |  | A state hash as per `#set_state` specifying properties to apply to all devices in the scene, overriding those configured in the scene.

### validate_color

#### Parameters

Key | Type | Mandatory | Default | Description
--- | --- | --- | --- | ---
`:color` | color | &#10004; |  | Color string you'd like to validate. See [color](https://api.developer.lifx.com/docs/colors) documentation on the LIFX website.


## Deviation from the API spec

Some API endpoints require a mandatory [`selector`](https://api.developer.lifx.com/docs/selectors) parameter, which defines which bulbs to apply your action to. This client will default the `selector` parameter to `'all'`, if no selector is provided.

```ruby
# which means that you can call
client.list_bulbs
# ...and receive a hash of all your bulbs back

# instead of having to explicitly specify you want all bulbs:
client.list_bulbs selector: 'all'
```

## Exceptions

If there is an error, `LifxApi` will raise an exception. The exception message will usually give a good indication of what went wrong, but you can also rescue the exception and access the request, response and decoded JSON objects, via the `request`, `response` and `data` methods.

## Development

Run `rake test` to run the tests and `rake console` to start an interactive pry console.

## TODO

* Validation of `:state` and `:array_of_states` is poor
* Validation of `:color` and `:array_of_colors` is poor
* Validation of endpoints is non-existent

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/cyclotron3k/lifx_api.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG
=== 0.2.0

* Add `fast` parameter to `LifxApi#set_state` and `LifxApi#set_states`
* Add `LifxApi#state_delta`, and deprecate `LifxApi#stage_delta`
* Add `LifxApi#move_effect`
* Add `LifxApi#morph_effect`
* Add `LifxApi#flame_effect`
* Add `LifxApi#effects_off`

=== 0.1.3

* Fix documentation and specification of `cycle` endpoint

=== 0.1.2

* Better validation of :selectors

* Unit tests for validators

* Fix some errors in the endpoints definition

=== 0.1.1

* Added endpoint documentation

=== 0.0.1

* Initial release


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "lifx_api"

require "minitest/autorun"


test/lifx_api_test.rb
require "test_helper"

class LifxApiTest < Minitest::Test
	def test_that_it_has_a_version_number
		refute_nil ::LifxApi::VERSION
	end

	def test_selector_validation
		lifx = LifxApi.new 'access_token'

		{
			'all' => true,
			'all:random' => true,
			'all|3,all:random' => true,
			'all|1-4' => true,
			'all|1|2|3|4' => true,
			'group:Lounge' => true,
			'group:Lounge:random' => true,
			'group_id:1c8de82b81f445e7cfaafae49b259c71' => true,
			'id:d3b2f2d97452' => true,
			'label:Bedroom,label:Office' => true,
			'label:Left Lamp' => true,
			'location:Home' => true,
			'location:Kitchen|1|5|9' => true,
			'location_id:1d6fe8ef0fde4c6d77b0012dc736662c' => true,
			'location_id:1d6fe8ef0fde4c6d77b0012dc736662c:random' => true,
			'scene_id:1d6fe8ef0fde4c6d77b0012dc736662c' => true,
			'banana' => false,
			'all,banana' => false,
			'all:random|3' => false,
		}.each do |selector, expected_result|

			assert_equal expected_result, lifx.send(:valid?, selector, :selector), "#{selector} failed :selector validation"

		end
	end

	def test_numeric_validation
		lifx = LifxApi.new 'access_token'

		{
			'1234' => true,
			1234 => true,
			'one' => false
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :numeric), "#{selector} failed :numeric validation"
		end
		# value.is_a?(Numeric) or value =~ /^[\d\.]+$/
	end

	def test_boolean_validation
		lifx = LifxApi.new 'access_token'

		{
			true => true,
			false => true,
			'true' => true,
			'false' => true,
			'maybe' => false,
			:nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :boolean), "#{selector} failed :boolean validation"
		end
		# ['true', 'false', true, false].include? value
	end

	def test_on_off_validation
		lifx = LifxApi.new 'access_token'

		{
			'on' => true,
			'off' => true,
			'Schrödinger' => false,
			nil => false,

		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :on_off), "#{selector} failed :on_off validation"
		end
		# ['on', 'off'].include? value
	end

	def test_hash_validation
		lifx = LifxApi.new 'access_token'

		{
			{} => true,
			[] => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :hash), "#{selector} failed :hash validation"
		end
		# value.is_a? Hash
	end

	def test_string_validation
		lifx = LifxApi.new 'access_token'

		{
			"Hello, World!" => true,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :string), "#{selector} failed :string validation"
		end
		# value.is_a? String
	end

	def test_uuid_validation
		lifx = LifxApi.new 'access_token'

		{
			'b854305d-81c9-4e56-b697-e47199ab9c7c' => true,
			'bananas' => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :uuid), "#{selector} failed :uuid validation"
		end
		# val
	end

	def test_ignore_array
		lifx = LifxApi.new 'access_token'

		{
			[] => true,
			['power'] => true,
			['brightness', 'duration', 'hue', 'infrared', 'intensity', 'kelvin', 'power', 'saturation'] => true,
			[''] => false,
			'power' => false,
			'brightness' => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :ignore_array), "#{selector} failed :ignore_array validation"
		end
	end

	def test_direction_validation
		lifx = LifxApi.new 'access_token'

		{
			'forward' => true,
			'backward' => true,
			'brightness' => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :direction), "#{selector} failed :direction validation"
		end
	end

	def test_state_validation
		lifx = LifxApi.new 'access_token'

		{
			{
				power: 'on',
				color: 'red',
				brightness: '1.0',
				duration: '1234',
				infrared: '1.0',
			} => true,
			{
				power: 'off',
				color: '#ff0000',
			} => true,
			{
				power: '50%',
			} => false,
			'brightness' => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :state), "#{selector} failed :state validation"
		end
	end

	def test_array_of_states_validation
		lifx = LifxApi.new 'access_token'

		{
			[{
				selector: 'all',
				power: 'on',
				color: 'red',
				brightness: '1.0',
				duration: '1234',
				infrared: '1.0',
			}] => true,
			[{
				selector: 'label:Kitchen',
				power: 'on',
			}, {
				selector: 'label:Playroom',
				infrared: 'off',
			}] => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :array_of_states), "#{selector} failed :array_of_states validation"
		end
	end

	def test_array_of_colors_validation
		lifx = LifxApi.new 'access_token'

		{
			['red', 'blue', 'green'] => true,
			['#ff0000'] => true,
			[123] => false,
			['red', 123] => false,
			{} => false,
			nil => false,
		}.each do |selector, expected_result|
			assert_equal expected_result, lifx.send(:valid?, selector, :array_of_colors), "#{selector} failed :array_of_colors validation"
		end
	end

	def test_validators_exist_for_all_data_types
		lifx = LifxApi.new 'access_token'

		LifxApi::ENDPOINTS.flat_map do |endpoint|
			endpoint.select { |k, _| /_params$/ === k.to_s }.values
		end.flat_map(&:values).map { |v| v[:type] }.sort.uniq.each do |value_format|

			# a missing test will raise an exception
			assert [true, false].include? lifx.send(:valid?, "", value_format)

		end

	end

end


lib/lifx_api.rb
require 'net/http'
require 'openssl'
require 'cgi'
require 'json'
require 'lifx_api/version'
require 'lifx_api/endpoints'
require 'lifx_api/error'

class LifxApi

	PROTOCOL = 'https'
	HOST     = 'api.lifx.com'
	DEBUG    = false

	def initialize(access_token)
		@access_token = access_token
		@base_uri = URI "#{PROTOCOL}://#{HOST}"

		@agent = Net::HTTP.new @base_uri.host, @base_uri.port
		@agent.use_ssl = PROTOCOL == 'https',
		@agent.keep_alive_timeout = 10
		@agent.set_debug_output $stdout if DEBUG
	end

	ENDPOINTS.each do |endpoint_spec|
		define_method endpoint_spec[:method_name], Proc.new { |params={}|
			if endpoint_spec[:deprecation_message]
				warn "[DEPRECATED] #{endpoint_spec[:deprecation_message]}"
			end
			parsed_params = parse_params endpoint_spec, params
			request = create_request endpoint_spec, parsed_params
			process_request request
		}
	end

	private

	def parse_params(endpoint_spec, params)
		[:path_params, :body_params, :query_params].each_with_object({}) do |field, parsed|
			next unless endpoint_spec[field]

			parsed[field] = endpoint_spec[field].each_with_object({}) do |(key, field_spec), clean|
				value = params[key] || field_spec[:default]
				raise "Missing #{key}" if field_spec[:required] and value.nil?
				next if value.nil?
				raise ArgumentError, "#{key} (#{value}) failed validation" unless valid? value, field_spec[:type]
				clean[key] = value
			end
		end
	end

	def valid?(value, value_format)
		case value_format
		when :selector
			value.split(',').all? do |selector|
				/^((label|id|(location|group)(_id)?|scene_id):.*|all)(:random|(\|[-\d]+)+)?$/ === selector
			end
		when :numeric, :brightness, :duration, :infrared
			value.is_a?(Numeric) or /^[\d\.]+$/ === value
		when :boolean
			['true', 'false', true, false].include? value
		when :on_off, :power
			['on', 'off'].include? value
		when :hash
			value.is_a? Hash
		when :string, :color
			value.is_a? String
		when :uuid
			value.is_a?(String) and /^[\da-f]{4}([\da-f]{4}-){4}[\da-f]{12}$/ === value
		when :ignore_array
			value.is_a?(Array) and (value - ['power', 'infrared', 'duration', 'intensity', 'hue', 'saturation', 'brightness', 'kelvin']).empty?
		when :direction
			value.is_a?(String) and ['forward', 'backward'].include? value
		when :array_of_states
			value.is_a?(Array) and value.count <= 50 and value.all? { |state| valid? state, :state }
		when :array_of_colors
			value.is_a?(Array) and value.all? { |state| valid? state, :color }
		when :state
			value.is_a?(Hash) and (value.keys - [:selector, :power, :color, :brightness, :duration, :infrared]).empty? and value.all? { |k, v| valid? v, k }
		else
			raise ArgumentError, "Don't know how to validate #{value_format}"
			true
		end
	end

	def create_request(endpoint_spec, params)
		uri = @base_uri.clone

		uri.path = if params.key? :path_params
			endpoint_spec[:path] % Hash[params[:path_params].map { |k, v| [k, CGI.escape(v).gsub('+', '%20')]}]
		else
			endpoint_spec[:path]
		end

		if params.key? :query_params
			uri.query = URI.encode_www_form params[:query_params]
		end

		request = case endpoint_spec[:http_method]
		when :get
			Net::HTTP::Get.new uri
		when :put
			Net::HTTP::Put.new uri
		when :post
			Net::HTTP::Post.new uri
		else
			raise NotImplementedError, "Invalid HTTP method: #{endpoint_spec[:http_method]}"
		end

		if params.key? :body_params
			request.set_form_data params[:body_params]
		end

		request
	end

	def process_request(request)
		request['Authorization'] = "Bearer #{@access_token}"

		puts "\n\n\e[1;32mDespatching request to #{request.path}\e[0m" if DEBUG

		@session = @agent.start unless @agent.active?
		response = @session.request request

		puts "\e[1;32mResponse received\e[0m" if DEBUG

		data = case response['content-type']
		when /application\/json/
			JSON.parse response.read_body, symbolize_names: true
		else
			raise "Don't know how to parse #{response['content-type']}"
		end

		unless (200..299).include? response.code.to_i
			raise LifxApi::Error.new request, response, data
		end

		data
	end
end


lib/lifx_api/endpoints.rb
class LifxApi
	ENDPOINTS = [{
		method_name: :list_lights,
		http_method: :get,
		path: '/v1/lights/%{selector}',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
	}, {
		method_name: :set_state,
		http_method: :put,
		path: '/v1/lights/%{selector}/state',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			power: {type: :string, description: 'The power state you want to set on the selector. Must be `on` or `off`.'},
			color: {type: :color, description: 'The color to set the light to.'},
			brightness: {type: :numeric, description: 'The brightness level from 0.0 to 1.0. Overrides any brightness set in color (if any).'},
			duration: {type: :numeric, default_description: '1.0', description: 'How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)'},
			infrared: {type: :numeric, description: 'The maximum brightness of the infrared channel from `0.0` to `1.0`.'},
			fast: {type: :boolean, default_description: 'false', description: 'Execute the query fast, without initial state checks and wait for no results.'},
		},
	}, {
		method_name: :set_states,
		http_method: :put,
		path: '/v1/lights/states',
		body_params: {
			states: {required: true, type: :array_of_states, description: 'Array of state hashes as per `#set_state`. No more than 50 entries.'},
			defaults: {type: :hash, description: 'Default values to use when not specified in each `states[]` hash.'},
			fast: {type: :boolean, default_description: 'false', description: 'Execute the query fast, without initial state checks and wait for no results.'},
		},
	}, {
		method_name: :state_delta,
		http_method: :post,
		path: '/v1/lights/%{selector}/state/delta',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'}
		},
		body_params: {
			power: {type: :on_off, description: 'The power state you want to set on the selector. Must be `on` or `off`.'},
			duration: {type: :numeric, default_description: '1.0', description: 'How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)'},
			infrared: {type: :numeric, description: 'The maximum brightness of the infrared channel.'},
			hue: {type: :numeric, description: 'Rotate the hue by this angle in degrees. Range: `-360.0` - `360.0` degrees.'},
			saturation: {type: :numeric, description: 'Change the saturation by this additive amount; the resulting saturation is clipped to `[0, 1]`.'},
			brightness: {type: :numeric, description: 'Change the brightness by this additive amount; the resulting brightness is clipped to `[0, 1]`.'},
			kelvin: {type: :numeric, description: 'Change the kelvin by this additive amount; the resulting kelvin is clipped to `[2500, 9000]`.'},
		}
	}, {
		method_name: :stage_delta,
		deprecation_message: '`#stage_delta` is deprecated, please use `#state_delta` instead',
		http_method: :post,
		path: '/v1/lights/%{selector}/state/delta',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'}
		},
		body_params: {
			power: {type: :on_off, description: 'The power state you want to set on the selector. Must be `on` or `off`.'},
			duration: {type: :numeric, default_description: '1.0', description: 'How long in seconds you want the power action to take. Range: `0.0` - `3155760000.0` (100 years)'},
			infrared: {type: :numeric, description: 'The maximum brightness of the infrared channel.'},
			hue: {type: :numeric, description: 'Rotate the hue by this angle in degrees. Range: `-360.0` - `360.0` degrees.'},
			saturation: {type: :numeric, description: 'Change the saturation by this additive amount; the resulting saturation is clipped to `[0, 1]`.'},
			brightness: {type: :numeric, description: 'Change the brightness by this additive amount; the resulting brightness is clipped to `[0, 1]`.'},
			kelvin: {type: :numeric, description: 'Change the kelvin by this additive amount; the resulting kelvin is clipped to `[2500, 9000]`.'},
		}
	}, {
		method_name: :toggle_power,
		http_method: :post,
		path: '/v1/lights/%{selector}/toggle',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			duration: {type: :numeric, default_description: '1.0', description: 'The time is seconds to spend perfoming the power toggle.'},
		},
	}, {
		method_name: :breathe_effect,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/breathe',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			color: {required: :true, type: :color, description: 'The color to use for the breathe effect.'},
			from_color:	{type: :string, default_description: 'current bulb color', description: 'The color to start the effect from. If this parameter is omitted then the color the bulb is currently set to is used instead.'},
			period: {type: :numeric, default_description: '1.0', description: 'The time in seconds for one cyles of the effect.'},
			cycles: {type: :numeric, default_description: '1.0', description: 'The number of times to repeat the effect.'},
			persist: {type: :boolean, default_description: 'false', description: 'If `false` set the light back to its previous value when effect ends, if true leave the last effect color.'},
			power_on: {type: :boolean, default_description: 'true', description: 'If `true`, turn the bulb on if it is not already on.'},
			peak: {type: :numeric, default_description: '0.5', description: 'Defines where in a period the target color is at its maximum. Minimum `0.0`, maximum `1.0`.'},
		}
	}, {
		method_name: :move_effect,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/move',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			direction: {type: :string, default_description: 'forward', description: 'Move direction, can be `forward` or `backward`.'},
			period: {type: :numeric, default_description: '1.0', description: 'The time in seconds for one cyles of the effect.'},
			cycles: {type: :numeric, default_description: 'infinite', description: 'The number of times to move the pattern across the device. Special cases are `0` to switch the effect off, and unspecified to continue indefinitely.'},
			power_on: {type: :boolean, default_description: 'true', description: 'Switch any selected device that is off to on before performing the effect.'},
		}
	}, {
		method_name: :morph_effect,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/morph',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			period: {type: :numeric, default_description: '5.0', description: 'This controls how quickly the morph runs. It is measured in seconds. A lower number means the animation is faster.'},
			duration: {type: :numeric, default_description: 'infinite', description: 'How long the animation lasts for in seconds. Not specifying a duration makes the animation never stop. Specifying `0` makes the animation stop. Note that there is a known bug where the tile remains in the animation once it has completed if duration is nonzero.'},
			palette: {type: :array_of_colors, default_description: '7 colours across the spectrum', description: 'You can control the colors in the animation by specifying a list of color specifiers. For example `["red", "hue:100 saturation:1"]`.'},
			power_on: {type: :boolean, default_description: 'true', description: 'Switch any selected device that is off to on before performing the effect.'},
		}
	}, {
		method_name: :flame_effect,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/flame',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			period: {type: :numeric, default_description: '5.0', description: 'This controls how quickly the flame runs. It is measured in seconds. A lower number means the animation is faster.'},
			duration: {type: :numeric, default_description: 'infinite', description: 'How long the animation lasts for in seconds. Not specifying a duration makes the animation never stop. Specifying `0` makes the animation stop. Note that there is a known bug where the tile remains in the animation once it has completed if duration is nonzero.'},
			power_on: {type: :boolean, default_description: 'true', description: 'Switch any selected device that is off to on before performing the effect.'},
		}
	}, {
		method_name: :pulse_effect,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/pulse',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			color: {required: true, type: :string, description: 'The color to use for the pulse effect.'},
			from_color: {type: :string, default_description: 'current bulb color', description: 'The color to start the effect from. If this parameter is omitted then the color the bulb is currently set to is used instead.'},
			period: {type: :numeric, default_description: '1.0', description: 'The time in seconds for one cyles of the effect.'},
			cycles: {type: :numeric, default_description: '1.0', description: 'The number of times to repeat the effect.'},
			persist: {type: :boolean, default_description: 'false', description: 'If `false` set the light back to its previous value when effect ends, if true leave the last effect color.'},
			power_on: {type: :boolean, default_description: 'true', description: 'If `true`, turn the bulb on if it is not already on.'},
		}
	}, {
		method_name: :effects_off,
		http_method: :post,
		path: '/v1/lights/%{selector}/effects/off',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			power_off: {type: :boolean, default_description: 'false', description: 'If `true`, the devices will also be turned off.'},
		}
	}, {
		method_name: :cycle,
		http_method: :post,
		path: '/v1/lights/%{selector}/cycle',
		path_params: {
			selector: {required: true, type: :selector, default: 'all'},
		},
		body_params: {
			states: {required: true, type: :array_of_states, description: 'Array of state hashes as per `#set_state`. Must have 2 to 5 entries.'},
			defaults: {type: :state, description: 'Default values to use when not specified in each `states[]` object.'},
			direction: {type: :direction, default_description: 'forward', description: 'Direction in which to cycle through the list. Can be `forward` or `backward`.'},
		}
	}, {
		method_name: :list_scenes,
		http_method: :get,
		path: '/v1/scenes',
	}, {
		method_name: :activate_scene,
		http_method: :put,
		path: '/v1/scenes/scene_id:%{scene_uuid}/activate',
		path_params: {
			scene_uuid: {required: true, type: :uuid, description: 'The UUID for the scene you wish to activate'},
		},
		body_params: {
			duration: {type: :numeric, default_description: '1.0', description: 'The time in seconds to spend performing the scene transition.'},
			ignore: {type: :ignore_array, description: 'Any of `power`, `infrared`, `duration`, `intensity`, `hue`, `saturation`, `brightness` or `kelvin`, specifying that these properties should not be changed on devices when applying the scene.'},
			overrides: {type: :state, description: 'A state hash as per `#set_state` specifying properties to apply to all devices in the scene, overriding those configured in the scene.'},
		}
	}, {
		method_name: :validate_color,
		http_method: :put,
		path: '/v1/color',
		query_params: {
			color: {type: :color, required: true, description: 'Color string you\'d like to validate'},
		}
	}]
end
